openforcefield/topology/molecule.py

```diff
@@ -4669,6 +4711,34 @@ def get_bond_between(self, i, j):
 
         raise NotBondedError("No bond between atom {} and {}".format(i, j))
 
+    @property
+    def rings(self):
+        if self._rings is None:
+            self._get_rings()
+        return self._rings
+
+    def _get_rings(self, toolkit_registry=GLOBAL_TOOLKIT_REGISTRY):
```

Comment:
Should we explicitly use the RDKit toolkit wrapper here to do the call as the method is only coded for that backend? So it would be similar to the `to_openeye` or `to_rdkit` methods, as with this current setup, this would also avoid the case of raising an error that the function could not be called if a user removes the RDKit toolkit wrapper from the `GLOBAL_TOOLKIT_REGISTRY `.

Reply:
Not sure; I like the idea of just letting `ToolkitWrapper.call` handle it, since maybe it will be implemented in another wrapper. But that's probably not on the horizon, and the error message doesn't really describe to the user that it's only available with RDKit installed:

```python3
In [1]: from openforcefield.topology import Molecule, Topology


In [2]: from openforcefield.utils.toolkits import *

In [3]: GLOBAL_TOOLKIT_REGISTRY.deregister_toolkit(RDKitToolkitWrapper)

In [4]: Molecule.from_smiles('c1ccc2ccccc2c1').n_rings
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-33b5c9295de5> in <module>
----> 1 Molecule.from_smiles('c1ccc2ccccc2c1').n_rings

~/software/openforcefield/openforcefield/topology/molecule.py in n_rings(self)
   3272     def n_rings(self):
   3273         """Return the number of rings found in the Molecule"""
-> 3274         return len(self.rings)
   3275
   3276     @property

~/software/openforcefield/openforcefield/topology/molecule.py in rings(self)
   4715     def rings(self):
   4716         if self._rings is None:
-> 4717             self._get_rings()
   4718         return self._rings
   4719

~/software/openforcefield/openforcefield/topology/molecule.py in _get_rings(self, toolkit_registry)
   4737
   4738         """
-> 4739         rings = toolkit_registry.call("find_rings", self)
   4740         self._rings = rings
   4741

~/software/openforcefield/openforcefield/utils/toolkits.py in call(self, method_name, raise_exception_types, *args, **kwargs)
   5023         for toolkit, error in errors:
   5024             msg += " {} {} : {}\n".format(toolkit, type(error), error)
-> 5025         raise ValueError(msg)
   5026
   5027     def __repr__(self):

ValueError: No registered toolkits can provide the capability "find_rings" for args "(Molecule with name '' and SMILES '[H]c1c(c(c2c(c(c(c(c2c1[H])[H])[H])[H])[H])[H])[H]',)" and kwargs "{}"
Available toolkits are: [ToolkitWrapper around OpenEye Toolkit version 2020.1.0, ToolkitWrapper around AmberTools version 20.0, ToolkitWrapper around Built-in Toolkit version None]


In [5]: GLOBAL_TOOLKIT_REGISTRY.register_toolkit(RDKitToolkitWrapper)

In [6]: Molecule.from_smiles('c1ccc2ccccc2c1').n_rings
Out[6]: 2
```

Reply:
Okay, after the next commit, these functions require RDKit. Here's what it looks like without (probably shouldn't be a `LicenseError`)

```
In [3]: mol.n_rings
---------------------------------------------------------------------------
LicenseError                              Traceback (most recent call last)
<ipython-input-3-3710dfea87bf> in <module>
----> 1 mol.n_rings

~/software/openforcefield/openforcefield/topology/molecule.py in n_rings(self)
   3277
   3278         """
-> 3279         return len(self.rings)
   3280
   3281     @property

~/software/openforcefield/openforcefield/topology/molecule.py in rings(self)
   4725         """
   4726         if self._rings is None:
-> 4727             self._get_rings()
   4728         return self._rings
   4729

~/software/openforcefield/openforcefield/utils/toolkits.py in wrapped_function(*args, **kwargs)
    206                         cls._toolkit_name
    207                     )
--> 208                     raise LicenseError(msg)
    209                 value = func(*args, **kwargs)
    210                 return value

LicenseError: This function requires the The RDKit toolkit
```

Reply:
Thanks for changing this I think this error is much better than the previous output and agree maybe the error type should change but at least the message is very clear.